clouds: support IBM VPC (SC-1351)

aciba90 opened the pull request:
## Proposed Commit Message
<!-- Include a proposed commit message because all PRs are squash merged -->

Desired merge method: Squash and use the following commit message:

```
clouds: add IBM VPC support

Add IBM VPC as a new cloud supporting Virtual Server Instances
and Bare Metal Servers.
```

## Additional Context
<!-- If relevant -->

[SDK reference](https://cloud.ibm.com/apidocs/vpc/latest?code=python#about-vpc-api)
[SC-1322](https://warthogs.atlassian.net/browse/SC-1322)
[SC-1351](https://warthogs.atlassian.net/browse/SC-1351)

## Test Steps
<!-- Please include any steps necessary to verify (and reproduce if
this is a bug fix) this change on a live deployed system,
including any necessary configuration files, user-data,
setup, and teardown. Scripts used may be attached directly to this PR. -->

```sh
# - Configure `pycloudlib.toml`.
# - Checkout https://github.com/canonical/cloud-init/pull/1915
# - Modify https://github.com/canonical/cloud-init/pull/1915/files#diff-7fd323c345163bbcf1eff443969951d75f9acc6bac3bbde442f143301e586621R5 to point to the last commit of this branch

CLOUD_INIT_PLATFORM=ibm tox -e integration-tests -- tests/integration_tests/modules/test_combined.py::TestCombined::test_netplan_permissions
```

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes
that apply. -->

- [x] Instance types (bare metal and dedicated hosts)
- [x] Constrain `ibm-cloud-sdk-core` for py36 only
- [x] Docs
- [x] Unit tests

[SC-1322]: https://warthogs.atlassian.net/browse/SC-1322?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[SC-1351]: https://warthogs.atlassian.net/browse/SC-1351?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

aciba90 commented:
Friendly ping just in case you, @hammerstefan, want to have a look.

TheRealFalcon commented on pycloudlib.toml.template:
```diff
@@ -45,6 +45,15 @@ project = ""  # glcoud config get-value project
 # private_key_path = ""  # Defaults to 'public_key_path' without the '.pub'
 # key_name = ""  # Defaults to your username if not set
 
+[ibm]
```
I think we might need to include VPC here.

E.g., if I run your cloud-init branch, I get `pycloudlib.ibm._util.IBMException: VPC not found: us-east-default-vpc` on every launch...even though I can see `us-east-default-vpc` if I log into the web console.

aciba90 replied:
Added the ability to set the VPC here.

The problem with that particular vpc (`us-east-default-vpc`) is that it is not associated to the resource group that you are probably using, otherwise it should have been detected.

TheRealFalcon reviewed (changes requested):
Bare metal out of the gate...nice!

I left a number of inline comments. Almost all of them are minor. The only major thing that I think we need to address is specifying the VPC via pycloudlib.toml.

Also, when using pyright on python 3.7 (and probably other versions), I do get a number of type errors related to arguments and return values from the IBM APIs. I'm assuming they've changed type signatures in a backwards compatible manner as it works on me on 3.6 and 3.11 too. I don't think we need to change anything, but it's just something to be aware of.

aciba90 commented on pycloudlib/ibm/_util.py:
```diff
@@ -0,0 +1,90 @@
+    return result
+
+
+def wait_until(
```
I think this is a great idea. Let's make it a follow-up issue: https://github.com/canonical/pycloudlib/issues/229

aciba90 commented on pycloudlib/ibm/_util.py:
```diff
@@ -0,0 +1,90 @@
+V = TypeVar("V")
+
+
+def get_all(
```
Yes, that would be a nice refactoring.
These functions grew organically while development, to end up pretty close.

aciba90 commented on pycloudlib/ibm/instance.py:
```diff
@@ -0,0 +1,735 @@
+        return raw_instance
+
+    @staticmethod
+    def _discover_floating_ip(client: VpcV1, instance: dict) -> Optional[dict]:
```
This function looks for a floating ip already associated to the instance and floating ips can only be associated to one thing, so I think it is safe to use.

TheRealFalcon replied:
Ahh, I didn't catch that it was only floating IPs assigned to the instance.

aciba90 commented on pycloudlib/ibm/instance.py:
```diff
@@ -0,0 +1,735 @@
+            raise IBMException("No subnet available")
+        return self._subnet.id
+
+    def delete(self) -> None:
```
This is only called in `examples/ibm.py`. VPC deletion is not called within pycloudlib.

I think if we support VPC creation, we have to support deletion, otherwise it would not be very useful for clients.

What do you suggest? Removing VPC creation and deletion, or keep it as is?

TheRealFalcon replied:
I think it makes sense to keep it. I was more worried about the possibility of automatically deleting something (e.g., like we do with floating ip on instance deletion). As long as it's an explicit operation, I think it's ok.

aciba90 commented on tox.ini:
```diff
@@ -16,7 +16,7 @@ deps =
 envdir = {[common]envdir}
 deps = {[common]deps}
-commands = {envpython} -m pytest {posargs:--cov pycloudlib}
+commands = {envpython} -m pytest --doctest-modules {posargs:--cov pycloudlib -- pycloudlib}
```
I see doctests as checked documentation / examples, not as a substitute for unit-testing.

TheRealFalcon replied:
That makes sense. I've seen people try to use it for unit testing and it quickly got out of control. As long as we don't try to do that, I think this makes sense.

aciba90 commented on pycloudlib/ibm/cloud.py:
```diff
@@ -0,0 +1,369 @@
+            return key["id"]
+
+        self._log.info("Creating SSH key: %s", self.key_pair.name)
+        return self._client.create_key(
```
That shouldn't fail in the first place, as `_get_or_create_key` should have fetched your key, let me try to reproduce and fix that.

aciba90 replied:
The problem was that:

1. you uploaded manually a key with a name
2. pycloudlib.toml without `key_name` specified
3. pycloudlib defaults to key_name = OS username and `~/.ssh/id_rsa.pub`
4. pycloudlib does not find a key associated to your username
5. pycloudlib uploads `~/ssh/id_rs.pub` with your username as name, but another key exists with the same fingerprint but a different name and IBM raises an exception.

I am not sure if it is worth pycloudlib trying to detect that situation and show a better error message. pycloudlib should either:

1) iterate over the uploaded fingerprints and compare them with the local one 
2) match on that IBM error msg and provide a better error message.

both approaches sound coupled and /or fragile. I would vote to keep it as is.

TheRealFalcon replied:
Yeah...I suppose this specific situation won't be very common. Keeping it as-is is fine.

aciba90 commented:
Thanks for the review, @TheRealFalcon!

> I left a number of inline comments. Almost all of them are minor. The only major thing that I think we need to address is specifying the VPC via pycloudlib.toml.

I have addressed the minor comments and added the ability to specify the VPC via pycloudlib.toml.

> Also, when using pyright on python 3.7 (and probably other versions), I do get a number of type errors related to arguments and return values from the IBM APIs. I'm assuming they've changed type signatures in a backwards compatible manner as it works on me on 3.6 and 3.11 too. I don't think we need to change anything, but it's just something to be aware of.

We could enable `mypy` in a follow-up task if we want to check type errors. What do you think?

TheRealFalcon reviewed (approved):
Really nice work here, @aciba90 , thanks!

I'll leave it up to you if you want to merge now or make any changes to get_first/get_all. 

> We could enable mypy in a follow-up task if we want to check type errors. What do you think?

We could, but it'll probably be trickier calling into these cloud APIs. I'm not going to rush into adding it :wink:

aciba90 commented:
Thanks, @TheRealFalcon!

> I'll leave it up to you if you want to merge now or make any changes to get_first/get_all.

I have refactored it in [c5bc867](https://github.com/canonical/pycloudlib/pull/225/commits/c5bc8672e29dcd09f91acbca19172dc35c12c53a) and I am going to proceed merging.
